notes.md

```diff
@@ -0,0 +1,255 @@
+3. Map `y_dense` onto `y`. Call this operation `(::Restructure)`.
+
+I'm going to define equivalence of output structurally -- two `AbstractArray`s are equal if
+1. they are primitives, and whatever notion of equality we have for them holds, or
+2. they are composite types, and each of their fields are equal under this definition.
+
+The reason for this notion of equality is that AD (as proposed above) treats concrete subtypes of AbstractArray no differently from any other composite type.
+
+Applying this recipe, a very literal implementation for the `rrule` of the equivalent function for `*` is something like the following:
+```julia
+function rrule(config::RuleConfig, ::typeof(*), A::AbstractMatrix, B::AbstractMatrix)
+
+    # Produce dense versions of A and B, and the pullbacks of this operation.
+    A_dense, destructure_A_pb = rrule(destructure, A)
+    B_dense, destructure_B_pb = rrule(destructure, B)
+
+    # Compute dense primal.
+    C_dense = A_dense * B_dense
+
+    # Compute structured primal without densifying to ensure that we get structured `C` back
+    # if that's what the primal would do.
+    C = A * B
+
+    # Construct pullback of Restructure. We generally need to extract some information from
+    # C in order to find the structured version.
+    _, restructure_C_pb = rrule_via_ad(config, Restructure(C), C_dense)
+
+    function my_mul_generic_pullback(C̄)
+
+        # Recover natural cotangent.
+        _, C̄_nat = restructure_C_pb(C̄)
+
+        # Compute pullback using natural cotangent of C.
+        Ā_nat = C̄_nat * B_dense'
+        B̄_nat = A_dense' * C̄_nat
+
+        # Transform natural cotangents w.r.t. A and B into structural (if non-primitive).
+        _, Ā = destructure_A_pb(Ā_nat)
+        _, B̄ = destructure_B_pb(B̄_nat)
+        return NoTangent(), Ā, B̄
+    end
+
+    # The output that we want is `C`, not `C_dense`, so return `C`.
+    return C, my_mul_generic_pullback
+end
+```
+All I've done is write out the rrule for differentiating through the equivalent function by hand.
+We'll optimise this implementation shortly to avoid e.g. having to densify primals, and computing the same function twice.
+`my_mul` in `examples.jl` verifies the correctness of the above implementation.
+
+
+`destructure` is quite straightforward to define -- essentially equivalent to `collect`. I'm confident that this is always going to be simple to define, because `collect` is always easy to define.
+
+`Restructure(C)(C_dense)` is a bit trickier. It's the function which takes an `Array` `C_dense` and transforms it into `C`. This feels like a slightly odd thing to do, since we already have `C`, but it's necessary to already know what `C` is in order to construct this function in general -- for example, over-parametrised matrices require this (see the `ScaledMatrix` example in the tests / examples). I'm _reasonably_ confident that this is always going to be possible to define, but I might have missed something.
```

Comment:
> I don't know how to do this for e.g. Woodbury, but I think you mentioned you thought about this and know how to do it?

The more I think about it, the less convinced I am that I know how to do this. I think this section on `Restructure` needs re-working / updating to reflect the increased level of uncertainty I have about how easy this is to do. For example, the `ScaledMatrix` example I mention is less straightforward than I originally thought.

> Similarly, two different structured matrices might map to the same dense matrix. In that case I suppose it doesn't matter which structured matrix we get for the correctness?

Yeah, I think that's totally fine. But is this a `destructure` thing rather than `Restructure`?